### PR TITLE
T-twr plus open edition support

### PR DESCRIPTION
**Pull request by nmuntyanov**
Hello.

THIS PR IS NOT PRODUCTION READY

I've found your project and interested in a lot.
I have T-TWR open edition VHF version 

So I took this https://github.com/OpenRTX/sa8x8-fw firmware, and wrote an abstraction layer for SA868 module
this library (https://github.com/OpenRTX/OpenRTX/blob/master/platform/drivers/baseband/AT1846S_SA8x8.cpp) was taken as a reference

I've tested APRS tracker, voice tx/rx

So if you interested in this, we can try to extend it and make it fully workable

### CODE REVIEW

**Comment by lyusupov** [comment]
@nmuntyanov you may consider to contribute your efforts _(if this PR merge will fail)_ into an alternative ESP32-S3 APRS project: https://github.com/Hamlabs/ArcticTracker-ESP32

**ArcticTracker-ESP32** is a mature project (4 years old). It uses a hardware which is quite similar to the T-TWR Plus.
The  **ArcticTracker-ESP32** author is friendly for porting the project on the T-TWR Plus as well:
1) https://github.com/Hamlabs/ArcticTracker-ESP32/issues/3
2) https://github.com/Hamlabs/ArcticTracker-ESP32?tab=readme-ov-file#future-work

image

**Comment by lyusupov** [comment]
@nmuntyanov 
use of exceptions (try...catch...)  is a bad practice for Arduino/PlatformIO compliant libraries/sketches.
https://forum.arduino.cc/t/try-catch/180063

Would you mind to alter your code with no use of exceptions ?

File: platformio.ini [hunk 1]
```diff
@@ -35,5 +35,12 @@ build_flags =
 	-DBOARD_HAS_PSRAM
 	-UARDUINO_USB_CDC_ON_BOOT
 	-mfix-esp32-psram-cache-issue
-	-DCORE_DEBUG_LEVEL=5
-;	-DLAST_BUILD_TIME=$UNIX_TIME
+
+[env:sa868_vhf_open_edition]
+build_flags= ${env.build_flags} -DSA868_OPEN_EDITION
+
+[env:sa868_vhf_open_edition_debug]
```
Comment (nmuntyanov) [inline platformio.ini:42]: @lyusupov maybe its better to do something like this?
separate build for each module?

**Comment by nmuntyanov** [comment]
@lyusupov 
> @nmuntyanov use of exceptions (try...catch...) is a bad practice for Arduino/PlatformIO compliant libraries/sketches. https://forum.arduino.cc/t/try-catch/180063
> 
> Would you mind to alter your code with no use of exceptions ?

try-catch blocks was removed

File: src/sa868_open_edition.cpp [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+#ifdef SA868_OPEN_EDITION
+
+bool SA868_WriteAT1846Sreg(HardwareSerial * SerialRF, uint8_t reg, uint16_t value)
+{
+    char str[200];
+    String result;
+    sprintf(str, "AT+POKE=%d,%d\r\n", reg, value);
+    if (!SA868_WaitResponse(SerialRF, str, &result)) {
+        ESP_LOGD("SA8x8", "Error: reg: %02X <- val: %02X", reg, value);
```
Comment (nmuntyanov) [inline src/sa868_open_edition.cpp:14]:
```suggestion
        ESP_LOGE("SA8x8", "Error: reg: %02X <- val: %02X", reg, value);
```

**Comment by lyusupov** [comment]
@nmuntyanov 

> try-catch blocks was removed

LGTM, thank you!

> separate build for each module?

IMHO, to split one firmware binary on two, _(radio specific)_ ones - it is a good start !

However, if you will eventually decide to support both NiceRF and OpenEdition drivers **within a single ESP32-S3 firmware binary** by having 2 instances of 'class SA868' and differentiate  them by executing Version()  call in runtime during the SA868 module probing sequence  - that would be perfect !

**Review by nakhonthai — APPROVED** [review]

**Review by nakhonthai — APPROVED** [review]

**Review by nakhonthai — APPROVED** [review]

**Comment by lyusupov** [comment]
@nmuntyanov 
I've pulled current 'master' of  the ESP32APRS_T-TWR repo and rebult -

```
======================================== [SUCCESS] Took 87.03 seconds ========================================

Environment                   Status    Duration
----------------------------  --------  ------------
sa868_vhf_open_edition        SUCCESS   00:02:49.842
sa868_vhf_open_edition_debug  SUCCESS   00:01:27.028
========================================= 2 succeeded in 00:04:16.870 =========================================

```

_My question is_: no PIO **build for SA868 with NiceRF firmware** anymore ?

**Comment by lyusupov** [comment]
@nmuntyanov 
I've made an attempt to use the firmware with Open Edition module.



<br>

### Issue No.1  
TRACKER function is enabled by default. It is worth to turn it back off.

image

<br>
<br>

### Issue No.2
Missing radio module info and unable to store settings.

image

**Comment by nmuntyanov** [comment]
@lyusupov 
it was draft pr, not production ready, I have added a notice in description.
I have just implemented driver for openEdition
no backward compatibility yet.

**Comment by nakhonthai** [comment]
Thank @nmuntyanov for code sa8x8.
I'm developing and testing.

messageImage_1706761421682
messageImage_1706761362047

**Pull request merged**